bytecomp/translmod.ml

```diff
@@ -249,6 +249,22 @@ let init_shape modl =
 
 type binding_status = Undefined | Inprogress | Defined
 
+let extract_unsafe_cycle id fv cycle_start =
+  let num_bindings = Array.length id in
+  let rec explore = function
+    | [] -> assert false
+    | (l,i) :: stack ->
+        if List.mem i l then
+          List.rev_map (Array.get id) (i::l)
+        else begin
+          let children =
+            List.filter
+              ( fun (_,j) -> Ident.Set.mem id.(j) fv.(i) )
+            @@ List.init num_bindings (fun j -> i::l, j ) in
+          explore (children @ stack) end
+  in
+  explore [[], cycle_start]
```

Comment:
I believe that the implementation is correct, but that a conceptually simpler approach (redoing less of the main loop's computations) would be possible: have `Inprogress of int` keep the parent of the node (in the end, parents depend on their children), and change `emit_binding i` to `emit_binding parent i` (the initial call is `emit_binding i i`). If `emit_binding parent i` reaches an `Inprogress _` node, you can construct the cycle by just recursively following `parent`'s parent until you hit `i` again.

Reply:
This works too. I have updated the code since it was indeed simpler (and I expanded the corresponding test to also test the cycle orientation).